src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedObjectCreationTests.cs

```diff
@@ -0,0 +1,3926 @@
+                Diagnostic(ErrorCode.ERR_TypelessNewNoTargetType, "new()").WithLocation(8, 22),
+                // (8,22): error CS8131: Deconstruct assignment requires an expression with a type on the right-hand-side.
+                //         (C _, C _) = new();
+                Diagnostic(ErrorCode.ERR_DeconstructRequiresExpression, "new()").WithLocation(8, 22)
+                );
+        }
+
+        [Fact]
+        public void TestBestType_NullCoalescing()
+        {
+            var source = @"
+using System;
+
+class C
+{
+    public static void Main()
+    {
+        C c = null;
+        Console.Write(c ?? new());
+    }
+}
+";
+
+            var comp = CreateCompilation(source, options: TestOptions.DebugExe).VerifyDiagnostics();
+
+            CompileAndVerify(comp, expectedOutput: "C");
+        }
+
+        [Fact]
+        public void TestBestType_Lambda()
+        {
+            var source = @"
+using System;
+
+class C
+{
+    public static void M<T>(Func<bool, T> f)
+    {
+        Console.Write(f(true));
+        Console.Write(f(false));
+    }
+    public static void Main()
+    {
+        M(b => { if (b) return new C(); else return new(); });
+    }
+}
+";
+
+            var comp = CreateCompilation(source, options: TestOptions.DebugExe).VerifyDiagnostics();
+
+            CompileAndVerify(comp, expectedOutput: "CC");
+        }
+
+        [Fact]
+        public void TestBestType_SwitchExpression()
+        {
+            var source = @"
+using System;
+
+class C
+{
+    public static void M<T>(Func<bool, T> f)
+    {
+        Console.Write(f(true));
+        Console.Write(f(false));
+    }
+    public static void Main()
+    {
+        var b = false; 
+        Console.Write((int)(b switch { true => 1, false => new() }));
+        b = true;
+        Console.Write((int)(b switch { true => 1, false => new() }));
+    }
+}
+";
+
+            var comp = CreateCompilation(source, options: TestOptions.DebugExe).VerifyDiagnostics();
+
+            CompileAndVerify(comp, expectedOutput: "01");
+        }
+
+        [Fact]
+        public void TestInSwitchExpression()
+        {
+            var source = @"
+using System;
+
+class C
+{
+    public static void M<T>(Func<bool, T> f)
+    {
+        Console.Write(f(true));
+        Console.Write(f(false));
+    }
+    public static void Main()
+    {
+        C x = 0 switch { _ => new() };
+        Console.Write(x);
+    }
+}
+";
+
+            var comp = CreateCompilation(source, options: TestOptions.DebugExe).VerifyDiagnostics();
+
+            CompileAndVerify(comp, expectedOutput: "C");
+        }
+
+        [Fact]
+        public void TestInNullCoalescingAssignment()
+        {
+            var source = @"
+using System;
+
+class C
+{
+    public static void M<T>(Func<bool, T> f)
+    {
+        Console.Write(f(true));
+        Console.Write(f(false));
+    }
+    public static void Main()
+    {
+        C x = null;
+        x ??= new();
```

Comment:
Please also test with nullable value types. Something like:
```cs
int? i = null;
i ??= new();
```

Based on other interactions, I'd expect that to be new int().